test/utils/MockERC1271VerifyingContract.sol

```diff
@@ -17,64 +19,60 @@ struct PermitDetails {
 }
 
 /// @title MockERC1271VerifyingContract
-/// @notice A mock contract that implements the ERC-1271 interface
-/// @dev This contract is used to test against our ERC-7739 implementation
+/// @notice A mock contract mirroring Permit2 which implements nested EIP-712 data structures
+/// @dev This contract is used to generate signatures for testing against our ERC-7739 implementation
 contract MockERC1271VerifyingContract is EIP712 {
-    string internal constant PERMIT_SINGLE_TYPE =
-        "PermitSingle(PermitDetails details,address spender,uint256 sigDeadline)PermitDetails(address token,uint160 amount,uint48 expiration,uint48 nonce)";
-    bytes32 internal constant PERMIT_SINGLE_TYPEHASH = keccak256(bytes(PERMIT_SINGLE_TYPE));
+    bytes32 public constant _PERMIT_SINGLE_TYPEHASH = keccak256(
+        "PermitSingle(PermitDetails details,address spender,uint256 sigDeadline)PermitDetails(address token,uint160 amount,uint48 expiration,uint48 nonce)"
+    );
 
-    string internal constant PERMIT_DETAILS_TYPE =
-        "PermitDetails(address token,uint160 amount,uint48 expiration,uint48 nonce)";
-    bytes32 internal constant PERMIT_DETAILS_TYPEHASH = keccak256(bytes(PERMIT_DETAILS_TYPE));
+    bytes32 public constant _PERMIT_DETAILS_TYPEHASH =
+        keccak256("PermitDetails(address token,uint160 amount,uint48 expiration,uint48 nonce)");
 
     constructor(string memory name, string memory version) EIP712(name, version) {}
 
+    /// @notice Public getter for the EIP-712 name
     function EIP712Name() external view returns (string memory) {
         return _EIP712Name();
     }
 
+    /// @notice Public getter for the EIP-712 version
     function EIP712Version() external view returns (string memory) {
         return _EIP712Version();
     }
 
+    /// @notice Public getter for the domain separator
     function domainSeparator() external view returns (bytes32) {
         return _domainSeparatorV4();
     }
 
-    /// return the full contents descriptor string
-    function contentsDescr() external pure returns (string memory) {
-        return PERMIT_SINGLE_TYPE;
+    /// @notice Per ERC-7739, use the explicit mode for content descriptor strings since the top level type
+    ///         PermitSingle is alphabetically ordered after PermitDetails
+    /// @dev return the full contents descriptor string in explicit mode
+    function contentsDescrExplicit() external pure returns (string memory) {
```

Comment:
should we test the implicit code path as well? also against ffi?

Reply:
yes good idea